pretyping/evarconv.ml

```diff
@@ -1196,20 +1196,19 @@ let check_problems_are_solved env evd =
   | (pbty,env,t1,t2) as pb::_ -> error_cannot_unify env evd pb t1 t2
   | _ -> ()
 
+exception MaxUndefined of Evar.t * evar_info
+
 let max_undefined_with_candidates evd =
-  (* If evar were ordered with highest index first, fold_undefined
-     would be going decreasingly and we could use fold_undefined to
-     find the undefined evar of maximum index (alternatively,
-     max_bindings from ocaml 3.12 could be used); instead we traverse
-     the whole map *)
-  let l = Evd.fold_undefined
-    (fun evk ev_info evars ->
-        match ev_info.evar_candidates with
-        | None -> evars
-        | Some l -> (evk,ev_info,l)::evars) evd [] in
-  match l with
-  | [] -> None
-  | a::l -> Some (List.last (a::l))
+  (* Rely on the decreasing order to find the undefined evar of maximum index *)
```

Comment:
This is a bit dry. Something like this would be better:
```
(* [fold_right] traverses the undefined map in decreasing order of indices. The evar with candidates of maximum index is thus the first evar with candidates found by a [fold_right] traversal. Taking advantage of this remark and exiting early has been found to have a significant impact on performance. *) 
```